### PR TITLE
Record: 11L Int5 + 6-Expert HedgeMixer + LeakyReLU(0.9)^2 + TTT (val_bpb=1.1105)

### PR DESCRIPTION
**Pull request by dttdrv**
## summary

11L 512d, 3.5x MLP with LeakyReLU(0.9)^2, 6-expert Hedge context mixer, GPTQ int5 + zstd-22, score-first TTT 4 epochs. 8xH100, 10min train + 10min eval.

val_bpb: 1.1105 (seed=1337)

## techniques

- 6-expert HedgeMixer: neural + unigram + bigram + trigram + 4-gram + entropy
- LeakyReLU(0.9)^2 activation
- GatedAttention + ValueResidual + XSA on all 11 layers
- Partial RoPE 16/64, LN Scale, BigramHash(8192), VE128
- EMA(0.997) + Tight SWA, Late QAT soft-round, CROWN-Q
- GPTQ int5, 3% pruning, zstd-22
- score-first TTT 4ep AdamW, byte-weighted, Polyak averaging

## metrics

- val_bpb: 1.1105 (TTT + HedgeMixer)
- pre-quant: 1.1342
- steps: 5684 at 101ms/step
- artifact: 15.95 MB
- train: 582s, eval: 570s

single seed. additional seeds pending compute.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: 11L Int5 + 6-Expert HedgeMixer + LeakyReLU(0.9)^2 + TTT (val_bpb=1.1105)

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)

--- ## Analysis ### 1. N-gram family bug check (CLEAN) `BigramHashEmbedding.bigram_hash` at line 725: ```python out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod ``` Hash key is formed from `t[..., 1:]` (current input token) XOR'd with `t[..., :-1]` (previous context token). This is `f(current_tok, prev_tok)` used to look up an embedding in `forward` called with `input_ids` (not `target_ids`). The target (`y`/`target_ids`) is **never XOR'd into the hash key**. No ILLEGAL n-gram family bug. ### 2. TTT pattern — score-first, is_last_chunk guard (CLEAN) `eval_val_sliding_ttt` implements a strict chunk-by-chunk pipeline: - **Phase 1 SCORE** (lines 1126–1173): chunk windows are scored under `torch.inference_mode()`. NLL accumulated into `loss_sum` / `token_count` / `byte_count`. - **Phase 2 TRAIN** (lines 1193–1251): training on `val_tokens[chunk_start:chunk_end]` happens only after scoring is complete for that chunk. Critical guard at line 1193: ```python if not is_last_chunk and ttt_epochs > 0 and elapsed_eval_time < time_limit: ``` `is_last_chunk = (ci == num_chunks - 1)` (line 1187). Training is **skipped for the final chunk**, matching the PR #1413 pattern. Every scored token was scored before any weight update could have seen it. No Pre-Quant TTT violation. ### 3. Mixer update (CLEAN) The `LogisticContextMixer.update()` call at line 1179 feeds only the already-scored chunk tokens into the n-gram count tables — it runs after Phase 1 scoring closes. Weights are updated via `update_weights()` inside the inference loop (line 1158) on already-computed `expert_nll`, not future tokens. ### 4. Quantization All-int5 (clip_range=15) with GPTQ calibration on training data. No scored-region SLOT concerns; the...

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern: each chunk scored under `torch.no_grad()` before `optimizer.step()`, with `is_last_chunk` guard preventing adaptation on the final scored chunk.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). TTT implementation follows the legal score-first discipline.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
